File: Algorithms/Graph_Algorithms/A_star/A_star.py
class Node:
    # Initialize the class
    def __init__(self, position:(), parent:()):
        self.position = position
        self.parent = parent
        self.g = 0 # Distance to start node
        self.h = 0 # Distance to goal node
        self.f = 0 # Total cost
    # Compare nodes
    def __eq__(self, other):
        return self.position == other.position
    # Sort nodes
    def __lt__(self, other):
         return self.f < other.f
    # Print node
    def __repr__(self):
        return ('({0},{1})'.format(self.position, self.f))
# Draw a grid
def draw_grid(map, width, height, spacing=2, **kwargs):
    for y in range(height):
        for x in range(width):
            print('%%-%ds' % spacing % draw_tile(map, (x, y), kwargs), end='')
        print()
# Draw a tile
def draw_tile(map, position, kwargs):
    
    # Get the map value
    value = map.get(position)
    # Check if we should print the path
    if 'path' in kwargs and position in kwargs['path']: value = '+'
    # Check if we should print start point
    if 'start' in kwargs and position == kwargs['start']: value = '@'
    # Check if we should print the goal point
    if 'goal' in kwargs and position == kwargs['goal']: value = '$'
    # Return a tile value
    return value 
# A* search
def astar_search(map, start, end):
    
    # Create lists for open nodes and closed nodes
    open = []
    closed = []
    # Create a start node and an goal node
    start_node = Node(start, None)
    goal_node = Node(end, None)
    # Add the start node
    open.append(start_node)
    
    # Loop until the open list is empty
    while len(open) > 0:
        # Sort the open list to get the node with the lowest cost first
        open.sort()
        # Get the node with the lowest cost
        current_node = open.pop(0)
        # Add the current node to the closed list
        closed.append(current_node)
        
        # Check if we have reached the goal, return the path
        if current_node == goal_node:
            path = []
            while current_node != start_node:
                path.append(current_node.position)
                current_node = current_node.parent
            #path.append(start) 
            # Return reversed path
            return path[::-1]
        # Unzip the current node position
        (x, y) = current_node.position
        # Get neighbors
        neighbors = [(x-1, y), (x+1, y), (x, y-1), (x, y+1)]
        # Loop neighbors
        for next in neighbors:
            # Get value from map
            map_value = map.get(next)
            # Check if the node is a wall
            if(map_value == '#'):
                continue
            # Create a neighbor node
            neighbor = Node(next, current_node)
            # Check if the neighbor is in the closed list
            if(neighbor in closed):
                continue
            # Generate heuristics (Manhattan distance)
            neighbor.g = abs(neighbor.position[0] - start_node.position[0]) + abs(neighbor.position[1] - start_node.position[1])
            neighbor.h = abs(neighbor.position[0] - goal_node.position[0]) + abs(neighbor.position[1] - goal_node.position[1])
            neighbor.f = neighbor.g + neighbor.h
            # Check if neighbor is in open list and if it has a lower f value
            if(add_to_open(open, neighbor) == True):
                # Everything is green, add neighbor to open list
                open.append(neighbor)
    # Return None, no path is found
    return None
# Check if a neighbor should be added to open list
def add_to_open(open, neighbor):
    for node in open:
        if (neighbor == node and neighbor.f >= node.f):
            return False
    return True
# The main entry point for this module
def main():
    # Get a map (grid)
    map = {}
    chars = ['c']
    start = None
    end = None
    width = 0
    height = 0
    # Open a file
    fp = open('data\\maze.in', 'r')
    
    # Loop until there is no more lines
    while len(chars) > 0:
        # Get chars in a line
        chars = [str(i) for i in fp.readline().strip()]
        # Calculate the width
        width = len(chars) if width == 0 else width
        # Add chars to map
        for x in range(len(chars)):
            map[(x, height)] = chars[x]
            if(chars[x] == '@'):
                start = (x, height)
            elif(chars[x] == '$'):
                end = (x, height)
        
        # Increase the height of the map
        if(len(chars) > 0):
            height += 1
    # Close the file pointer
    fp.close()
    # Find the closest path from start(@) to end($)
    path = astar_search(map, start, end)
    print()
    print(path)
    print()
    draw_grid(map, width, height, spacing=1, path=path, start=start, goal=end)
    print()
    print('Steps to goal: {0}'.format(len(path)))
    print()
# Tell python to run main method
if __name__ == "__main__": main()

File: Algorithms/Graph_Algorithms/A_Star/A_star.ts
/* GraphNode clas for A* Pathfinding 
 parent is parent of the current Node
 position is current psition of the Node in the maze
 g is cost from start to current Node
 h is heuristic based estimated cost for current Node to end Node
 f is total cost of present node: i.e. : f = g + h
*/
class GraphNode {
    public g: number;
    public h: number;
    public f: number;
    public position: Position;
    public parent: GraphNode | undefined;
    constructor(position: Position);
    constructor(position: Position, parent: GraphNode);
    constructor(position: Position, parent?: GraphNode) {
        this.g = 0;
        this.h = 0;
        this.f = 0;
        this.position = position;
        this.parent = parent;
    }

    isEqual(other: Position): boolean {
        return this.position.x == other.x && this.position.y == other.y;
    }
}

// Check if node is within the graph
const isValid = (grid: number[][], row: number, col: number): boolean => {
    return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length;
};

// Check if node is not blocked node
const isUnBlocked = (grid: number[][], row: number, col: number) => {
    return grid[row][col] == 1;
};

const reachedDestination = (src: Position, dst: Position) => {
    return src.x == dst.x && src.y == dst.y;
};

const inList = (child: GraphNode, list: GraphNode[]): null | GraphNode => {
    for (let node of list) {
        if (node.isEqual(child.position)) {
            return node;
        }
    }
    return null;
};

const aStarSearch = (grid: number[][], src: Position, dst: Position) => {
    // If source is invalid
    if (!isValid(grid, src.x, src.y)) {
        console.log("Source is invalid");
        return;
    }
    // If destionation is invalid
    if (!isValid(grid, dst.x, dst.y)) {
        console.log("Destination is invalid");
    }

    // Check if the source and destination node code is blocked
    if (!isUnBlocked(grid, src.x, src.y) || !isUnBlocked(grid, dst.x, dst.y)) {
        console.log("Source or the destination is blocked");
    }

    //Check the destination node is the same as the source cell
    if (reachedDestination(src, dst)) {
        console.log("Destination is already reached ");
    }

    // Initialize startNode and endNode
    let startNode = new GraphNode(src);
    let endNode = new GraphNode(dst);

    console.log(startNode);
    console.log(endNode);

    // Initialize both open and closed list
    let openList: GraphNode[] = [];
    let closedList: GraphNode[] = [];

    // Add the start node to open list
    openList.push(startNode);

    while (openList.length > 0) {
        // Get the current node
        let currentNode = openList[0];
        let currentIndex = 0;
        for (let i = 0; i < openList.length; i++) {
            if (openList[i].f < currentNode.f) {
                currentNode = openList[i];
                currentIndex = i;
            }
        }

        // Pop current off open list, add to closed list
        openList.splice(currentIndex, 1);
        closedList.push(currentNode);

        // if the destination node is reached
        if (currentNode.isEqual(endNode.position)) {
            let path = [];
            let current: GraphNode | undefined = currentNode;
            while (current != undefined) {
                path.push(current.position);
                current = current.parent;
            }
            return path.reverse();
        }

        // Generate children
        let children = [];
        let neighbours = [
            new Position(0, -1),
            new Position(0, 1),
            new Position(-1, 0),
            new Position(1, 0),
            new Position(-1, -1),
            new Position(-1, 1),
            new Position(1, -1),
            new Position(1, 1),
        ];

        for (let position of neighbours) {
            // Get node postion
            let nodePosition = new Position(
                currentNode.position.x + position.x,
                currentNode.position.y + position.y
            );

            // Make sure within the range
            if (!isValid(grid, nodePosition.x, nodePosition.y)) {
                continue;
            }

            if (!isUnBlocked(grid, nodePosition.x, nodePosition.y)) {
                continue;
            }

            let newNode = new GraphNode(nodePosition, currentNode);
            children.push(newNode);

            // Loop through children
            for (let child of children) {
                if (inList(child, closedList)) {
                    continue;
                }

                child.g = currentNode.g + 1;
                child.h =
                    (child.position.x - endNode.position.x) ** 2 +
                    (child.position.y - endNode.position.y) ** 2;
                child.f = child.g + child.h;

                let openNode = inList(child, openList);
                if (openNode && child.g > openNode.g) {
                    continue;
                }

                // Add the child to the open list
                openList.push(child);
            }
        }
    }
};

class Position {
    constructor(public x: number, public y: number) {}
}

// 1--> The cell is not blocked
// 0--> The cell is blocked
let grid: number[][] = [
    [1, 0, 1, 1, 1, 1, 0, 1, 1, 1],
    [1, 1, 1, 0, 1, 1, 1, 0, 1, 1],
    [1, 1, 1, 0, 1, 1, 0, 1, 0, 1],
    [0, 0, 1, 0, 1, 0, 0, 0, 0, 1],
    [1, 1, 1, 0, 1, 1, 1, 0, 1, 0],
    [1, 0, 1, 1, 1, 1, 0, 1, 0, 0],
    [1, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    [1, 0, 1, 1, 1, 1, 0, 1, 1, 1],
    [1, 1, 1, 0, 0, 0, 1, 0, 0, 1],
];

let src = new Position(8, 0);

let dst = new Position(0, 0);

let path = aStarSearch(grid, src, dst);
console.log(path);
